### DISCUSSION TITLE
Deposit function in new smart contract (Coin vs FungibleAsset)

### DISCUSSION THREAD

**Discussion by artur-abliazimov** [aptos-labs/aptos-developer-discussions#561] [Questions] [answered]
### Describe your question in detail.

Dear Devs!

I'm writing a smart contract that accepts different assets, including APT. I have a deposit function. Considering the migration of Coin to FungibleAsset, what should I do?

 - Write a deposit function only for FungibleAsset?
 - Write a deposit function only for Coin?
 - Write two deposit functions: one for Coin and another for FungibleAsset?

The function for FungibleAsset:
```
public entry fun deposit(sender: &signer, token_metadata: Object<Metadata>, amount: u64)
```

The function for Coin:
```
public entry fun deposit<CoinType>(sender: &signer, amount: u64)
```

For reference:
https://medium.com/aptoslabs/live-on-aptos-mainnet-coin-to-fungible-asset-migration-58eacaeaf7f7

### Which operating system are you using?

macOS

### Which SDK or tool are you using? (if any)

Aptos CLI

**Comment by Alexzander-Stone** [comment, marked as answer]
A fourth option is to write a single entry function that uses conditionals to check which flow to walk through. 😌

The Fungible Asset flow will be your standard, while the Coin flow will have an additional step of converting the coin to fungible asset using `coin_to_fungible_asset` ([reference](https://aptos.dev/en/build/smart-contracts/reference?page=aptos-framework%2Fdoc%2Fcoin.md#0x1_coin_coin_to_fungible_asset)) tacked onto the beginning.

**Reply by artur-abliazimov** [reply]
It worked! Thanks!

**Reply by kshitijc1506** [reply]
You should consider the implications of withdrawing to accounts in FA form, if they were deposited as coins

For instance, until the APT migration is complete, it makes sense to withdraw the APT as Coin (at least on mainnet) -- for CEX compatibility, DEX pools, other DeFi protocols, and general ecosystem support

You can use `coin::withdraw` -- it converts FA back to Coin as the funds leave the contract

**Reply by artur-abliazimov** [reply]
Yeah, this makes sense. Thanks!
